### ISSUE TITLE
unnecessary_unwrap_unchecked false positive

### ISSUE THREAD

**Issue by mathisbot** [rust-lang/rust-clippy#17349]
### Summary

The lint unnecessary_unwrap_unchecked triggers even inside the definition of the functions that defines the unchecked conversion.

### Lint Name

unnecessary_unwrap_unchecked

### Reproducer

I tried this code: https://rust.godbolt.org/z/PfrGnoYjW

```rust
pub enum Foo {
    A,
    B,
}

impl Foo {
    pub const fn as_u8(self) -> u8 {
        match self {
            Self::A => 0,
            Self::B => 1,
        }
    }

    pub const fn from_u8(value: u8) -> Option<Self> {
        match value {
            0 => Some(Self::A),
            1 => Some(Self::B),
            _ => None,
        }
    }

    /// # Safety
    ///
    /// The value must be either 0 or 1
    pub const unsafe fn from_u8_unchecked(value: u8) -> Self {
        unsafe { Self::from_u8(value).unwrap_unchecked() }
    }
}

```

I saw this happen:

```
warning: usage of `unwrap_unchecked` when an `_unchecked` variant of the associated function exists
  --> <source>:26:18
   |
26 |         unsafe { Self::from_u8(value).unwrap_unchecked() }
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_unwrap_unchecked
   = note: `#[warn(clippy::unnecessary_unwrap_unchecked)]` on by default
help: use `Foo::from_u8_unchecked` instead, and remove the call to `.unwrap_unchecked()`
   |
26 -         unsafe { Self::from_u8(value).unwrap_unchecked() }
26 +         unsafe { Self::from_u8_unchecked(value) }
```

I expected to see this happen:

No lint

### Version

```text
latest nightly (see godbolt)
```
